x/group/server/invariants.go

```diff
@@ -0,0 +1,97 @@
+package server
+
+import (
+	"math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/regen-network/regen-ledger/orm"
+	"github.com/regen-network/regen-ledger/x/group"
+)
+
+func (s serverImpl) RegisterInvariants(ir sdk.InvariantRegistry) {
+	ir.RegisterRoute(group.ModuleName, "Tally-Votes", s.tallyVotesInvariant())
+}
+
+func (s serverImpl) tallyVotesInvariant() sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		if ctx.BlockHeight()-1 < 0 {
+			return sdk.FormatInvariant(group.ModuleName, "Tally-Votes", "Not enough blocks to perform TallyVotesInvariant"), false
+		}
+		prevCtx, _ := ctx.CacheContext()
+		prevCtx = prevCtx.WithBlockHeight(ctx.BlockHeight() - 1)
+		msg, broken, err := tallyVotesInvariant(ctx, prevCtx, s.proposalTable)
+		if err != nil {
+			panic(err)
+		}
+		return sdk.FormatInvariant(group.ModuleName, "Tally-Votes", msg), broken
+	}
+}
+
+func tallyVotesInvariant(ctx sdk.Context, prevCtx sdk.Context, proposalTable orm.AutoUInt64Table) (string, bool, error) {
+
+	var msg string
+	var broken bool
+
+	prevIt, err := proposalTable.PrefixScan(prevCtx, 1, math.MaxUint64)
+	if err != nil {
+		return msg, broken, err
+	}
+	curIt, err := proposalTable.PrefixScan(ctx, 1, math.MaxUint64)
+	if err != nil {
+		return msg, broken, err
+	}
+
+	var curProposals []*group.Proposal
+	_, err = orm.ReadAll(curIt, &curProposals)
+	if err != nil {
+		return msg, broken, err
+	}
+
+	var prevProposals []*group.Proposal
+	_, err = orm.ReadAll(prevIt, &prevProposals)
+	if err != nil {
+		return msg, broken, err
+	}
+
+	for i := 0; i < len(prevProposals) && i < len(curProposals); i++ {
+		prevYesCount, err := prevProposals[i].VoteState.GetYesCount()
+		if err != nil {
+			return msg, broken, err
+		}
+		curYesCount, err := curProposals[i].VoteState.GetYesCount()
```

Comment:
Are we sure that `prevProposals[i] == curProposals[i]` always holds? (I guess it depends on the implementation of orm.ReadAll). Maybe an invariant `prevProposals[i].ProposalID == curProposals[i].ProposalID` could be useful, or at least a comment about this.

Also, we can replace `for i := 0; i < len(prevProposals) && i < len(curProposals)` by `for i := 0; i < len(prevProposals)` right? Since there's no proposal deletion mechanism.

Reply:
Thanks @amaurym 
I too think that we can replace the loop